src/extension/rpc.ts

```diff
@@ -0,0 +1,191 @@
+import { NoInfer, SafeAny } from "../types";
+import { MessageAdapter } from "./messageAdapters";
+import {
+  ApolloClientDevtoolsRPCMessage,
+  MessageType,
+  isRPCMessage,
+} from "./messages";
+
+type RPCParams = Record<string, unknown>;
+
+type RPCRequestMessageFormat = {
+  type: string;
+  params: RPCParams;
+};
+
+type RPCResponseMessageFormat =
+  | { sourceId: number; result: unknown }
+  | { sourceId: number; error: unknown };
+
+export type RPCMessage<
+  Name extends string,
+  Params extends RPCParams,
+  ReturnType,
+> = {
+  __name: Name;
+  __params: Params;
+  __returnType: ReturnType;
+};
+
+export interface RpcClient<
+  Messages extends RPCMessage<string, RPCParams, SafeAny>,
+> {
+  request: <TName extends Messages["__name"]>(
+    name: TName,
+    params: Extract<Messages, { __name: TName }>["__params"],
+    options?: { timeoutMs?: number }
+  ) => Promise<Extract<Messages, { __name: TName }>["__returnType"]>;
+}
+
+let nextMessageId = 0;
```

Comment:
I might consider adding something like [cuid2](https://github.com/paralleldrive/cuid2) to create the message ids. These are extremely subject to collisions and different areas of the devtools will initialize this value several times (i.e. its not shared). It _shouldn't_ be a problem for what we are using it for, but might be good to ensure `id` integrity regardless.

Reply:
We could also stick to browser apis to avoid pulling in another dependency:
https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID or, as I'm not sure if we're always in a "secure context", https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues

Reply:
Updated in [cc691ac](https://github.com/apollographql/apollo-client-devtools/pull/1256/commits/cc691ace76eaad32b2488d8eca9bd522b2d1da1a) by using `crypto.getRandomValues()`